[WIP] Show page for blocks in the future

SeaLightHouse opened the pull request:
Issue: https://github.com/decred/dcrdata/issues/372
When data about a  block which is yet to be mined is requested  then previously it showed a generalized error message . I have modified the error message so that if the block is yet to be mined it will show the estimated time for the mining of that block . I have added code in websocket handler for newblock event to show the block page as soon as it is ready

SeaLightHouse commented:
in file extras.tmpl the new code is inside an existing eventhandler . Is it good to create a new eventhandler just for this code ?

SeaLightHouse commented:
https://drive.google.com/file/d/1_doHb6WJnKf55R4M3sk2qwsiMk9KbE-n/view

chappjc commented on views/extras.tmpl:
```diff
@@ -278,6 +278,10 @@
             }
+            //handling statuspage for future block 
+            if($("#futureblock").length ){
+                location.reload(true);
```
please try to reload/refresh with turbolinks so as not to break the open websocket connection
 https://github.com/turbolinks/turbolinks/issues/329#issuecomment-341695200

chappjc commented on explorer/explorermiddleware.go:
```diff
@@ -48,7 +48,7 @@ func (exp *explorerUI) BlockHashPathOrIndexCtx(next http.Handler) http.Handler {
 				expectedTime := time.Duration(height-maxHeight) * exp.ChainParams.TargetTimePerBlock
 				message := fmt.Sprintf("This block is expected to arrive in approximately in %v. ", expectedTime)
-				exp.StatusPage(w, defaultErrorCode, message, NotFoundStatusType)
+				exp.StatusPage(w, "Future Block", message, NotFoundStatusType)
```
please define a new error message alongside the others in https://github.com/decred/dcrdata/blob/0dfef89dbc01b471063d1f4e41a88bc24ae6421e/explorer/explorerroutes.go#L30

SeaLightHouse replied:
OK , I created this const : unminedBlock

const (
        defaultErrorCode    = "Something went wrong..."
        defaultErrorMessage = "Try refreshing this page... it usually fixes things"
        fullModeRequired    = "full-functionality mode required for this page"
        unminedBlock = "Future Block"
)
So this issue is fixed

SeaLightHouse commented:
> Thanks for the edits. Please test to be sure the `Turbolinks.visit(window.location)` solution works as intended, keeping the websocket connection opened and the page updated without a full reload.
> 
> Also, please make git commits descriptive of the edits themselves rather than just saying they are addressing reviews. Thanks.

I have  tested 2 cases again today . 
1. Transitioning from one bock back to valid block
2. Transitioning from more than one future block to lesser one .
For case 1 it  is not disconnected. For case 2 it gets disconnected. But the web page is ok in bot pages . Let me check that .
Thanks

chappjc commented:
Did the status OK change fix the issue?  Ready for more testing on my end?

SeaLightHouse commented:
Yes, I have tested it , and to be precise , the http statusOK fixed the issue of websocket disconnecting and reconnected . Please test on your end too.

SeaLightHouse commented:
This line is in /block page :
	w.Header().Set("Turbolinks-Location", r.URL.RequestURI())
I guess , that is not needed . I have not added  that status page

SeaLightHouse commented:
I will work to remove the conflict today

chappjc commented:
Lemme know if you need help with the rebase and resolve.

SeaLightHouse commented:
Sure , let me try first.

chappjc commented:
So that didn't work.  What did you merge?  There should be no need to merge anything.  squash and rebase are not merge operations.

SeaLightHouse commented:
OK, somehow that did not work. I am going through git manual as it is new to me . Give me  a day . I will get it straightened up ready for merge .

chappjc commented:
What branches are you merging to and from? I'm lost as to what's going on here.

SeaLightHouse commented:
Ok , I will be careful next time . Shall I close this pr and make a new pr ?

chappjc commented:
If you need to make a new one OK.  If you can fix up this branch, that's better, but not worth holding up the show for.

SeaLightHouse commented:
OK, I am making a new pr for this issue . I will submit the pr request by today night. I think it is not correct to waste time on aspects of git , I can learn more on git so that it helps next.

chappjc commented:
Superseded by https://github.com/decred/dcrdata/pull/703.  @SeaLightHouse please review that PR.
